### PR TITLE
CBL-5793: Add missing c4 APIs for c4Index.

### PR DESCRIPTION
**Pull request by callumbirks**
`getName` and `getCollection` exist in the C++ API, but not the C API.
This PR rectifies that.

### CODE REVIEW

File: C/c4CAPI.cc [hunk 1]
```diff
@@ -1059,6 +1060,14 @@ C4IndexUpdater* C4NULLABLE c4index_beginUpdate(C4Index* index, size_t limit, C4E
     return tryCatch<C4IndexUpdater*>(outError, [&] { return index->beginUpdate(limit).detach(); });
 }
 
+C4SliceResult c4index_getName(C4Index* index) noexcept {
```
Comment (jianminzhao) [inline C/c4CAPI.cc:1063]: should handle the exception

File: C/c4CAPI.cc [hunk 2]
```diff
@@ -1059,6 +1060,14 @@ C4IndexUpdater* C4NULLABLE c4index_beginUpdate(C4Index* index, size_t limit, C4E
...
+    return C4SliceResult(index->getName());
+}
+
+C4Collection* c4index_getCollection(C4Index* index) noexcept {
```
Comment (jianminzhao) [inline C/c4CAPI.cc:1067]: handle exception

**Review by jianminzhao — APPROVED** [review]

File: C/c4CAPI.cc (hunk 1 above)
Comment (callumbirks) [reply C/c4CAPI.cc:1063]: What exception?

File: C/c4CAPI.cc (hunk 2 above)
Comment (callumbirks) [reply C/c4CAPI.cc:1067]: What exception?

File: C/c4CAPI.cc (hunk 1 above)
Comment (jianminzhao) [reply C/c4CAPI.cc:1063]: noexcept means the app will be terminated should exception popped to here. We do something like, tryCatch<C4 ...>

Comment (callumbirks) [reply C/c4CAPI.cc:1063]: Nevermind 😅

File: C/include/c4Index.h [hunk 3]
```diff
@@ -136,6 +138,16 @@ CBL_CORE_API bool c4coll_isIndexTrained(C4Collection* collection, C4String name,
     @return  A new `C4IndexUpdater` reference, or NULL if there's nothing to update. */
 CBL_CORE_API C4IndexUpdater* C4NULLABLE c4index_beginUpdate(C4Index* index, size_t limit, C4Error* outError) C4API;
 
+/**
+ * Return the name of this index.
+ **/
+CBL_CORE_API C4SliceResult c4index_getName(C4Index* index) C4API;
```
Comment (snej) [inline C/include/c4Index.h:144]: The underlying c4Index::getName returns a `slice`, so this result type should be `C4String`, not `C4SliceResult`. (I.e. there's no need to copy the name and force the caller to release it.)

File: C/c4CAPI.cc [hunk 4]
```diff
@@ -1059,6 +1060,14 @@ C4IndexUpdater* C4NULLABLE c4index_beginUpdate(C4Index* index, size_t limit, C4E
     return tryCatch<C4IndexUpdater*>(outError, [&] { return index->beginUpdate(limit).detach(); });
 }
 
+C4SliceResult c4index_getName(C4Index* index) noexcept {
+    return tryCatch<C4SliceResult>(nullptr, [&]{ return C4SliceResult(index->getName()); });
+}
+
+C4Collection* c4index_getCollection(C4Index* index) noexcept {
+    return tryCatch<C4Collection*>(nullptr, [&]{ return index->getCollection(); });
```
Comment (snej) [inline C/c4CAPI.cc:1068]: c4Index::getCollection is a trivial getter that just returns a member, so there's no need to wrap a try/catch around this. (If you like, you could add `noexcept` to the C++ method declaration.)

**Review by snej — CHANGES_REQUESTED** [review]

File: C/include/c4Index.h (hunk 3 above)
Comment (callumbirks) [reply C/include/c4Index.h:144]: Okay thanks :)

File: C/c4CAPI.cc (hunk 4 above)
Comment (callumbirks) [reply C/c4CAPI.cc:1068]: Oh okay I thought we might need it incase of null pointer access exception. But if not, I'll remove it

Comment (snej) [reply C/c4CAPI.cc]: Null pointer access is a signal/crash, not an exception, in C/C++.

File: C/include/c4Index.h [hunk 5]
```diff
@@ -11,6 +11,8 @@
 //
 
 #pragma once
+#include "c4Base.h"
+#include "c4Compat.h"
```
Comment (snej) [inline C/include/c4Index.h:15]: That's not necessary. Those headers are transitively included by c4IndexTypes.h, and none of the other headers (to my knowledge) redundantly include them.

File: C/c4CAPI.cc [hunk 6]
```diff
@@ -10,6 +10,7 @@
 // the file licenses/APL2.txt.
 //
 
+#include "c4Base.h"
```
Comment (snej) [inline C/c4CAPI.cc:13]: That's not necessary. That header is transitively included by the other API headers (e.g. c4BlobStore.hh), and none of the other headers (to my knowledge) redundantly include them.

**Review by snej — CHANGES_REQUESTED** [review]

File: C/c4CAPI.cc (hunk 6 above)
Comment (jianminzhao) [reply C/c4CAPI.cc]: removed.

File: C/include/c4Index.h (hunk 5 above)
Comment (jianminzhao) [reply C/include/c4Index.h]: removed

**Review by snej — APPROVED** [review]

**Pull request merged**